### PR TITLE
fix: Refactor `LayersMap` to correct old strange code behavior

### LINKED ISSUE #1420: Build failed with "removing whiteout sys/.wh..wh..opq: unlinkat /sys/.wh..opq: read-only file system"
**Actual behavior**
The kaniko build failed while unpacking and whiteouting file systems. (Debug log is available in Additional Information section)
The build is working if doing with "docker build".

> INFO[0002] Unpacking rootfs as cmd COPY test.lic test.lic requires it.
> error building image: error building stage: failed to get filesystem from image: removing whiteout sys/.wh..wh..opq: unlinkat /sys/.wh..opq: read-only file system

**Expected behavior**
Kaniko build should be successful if there is COPY command in Dockerfile.

**To Reproduce**
Run the below kaniko build on Docker
> docker run -v "$HOME"/.config/gcloud:/root/.config/gcloud -v /home/ec2-user:/workspace/ gcr.io/kaniko-project/executor:latest --context dir:///workspace/ --no-push --verbosity debug

**Additional Information**
 - Dockerfile 
> FROM pingidentity/pingfederate
COPY test.lic test.lic

 - Build Context
> ls -l
total 8
-rw-rw-r-- 1 ec2-user ec2-user 54 Sep  8 10:28 Dockerfile
-rw-rw-r-- 1 ec2-user ec2-user  5 Sep  8 10:31 test.lic

 - Kaniko Image (fully qualified with digest)
 gcr.io/kaniko-project/executor:latest

 - Full Debug Log
> docker run -v "$HOME"/.config/gcloud:/root/.config/gcloud -v /home/ec2-user:/workspace/ gcr.io/kaniko-project/executor:latest --context dir:///workspace/ --no-push --verbosity debug
DEBU[0000] Getting source context from dir:///workspace/
DEBU[0000] Build context located at /workspace/
DEBU[0000] Copying file /workspace/Dockerfile to /kaniko/Dockerfile
DEBU[0000] Skip resolving path /kaniko/Dockerfile
DEBU[0000] Skip resolving path /workspace/
DEBU[0000] Skip resolving path /cache
DEBU[0000] Skip resolving path
DEBU[0000] Skip resolving path
DEBU[0000] Skip resolving path
DEBU[0000] Built stage name to index map: map[]
INFO[0000] Retrieving image manifest pingidentity/pingfederate
INFO[0000] Retrieving image pingidentity/pingfederate
DEBU[0000] No file found for cache key sha256:9cbf21e56fa11e140e68f95fb9d855d7b5a6468e1ccebb647c5078d169eed9d2 stat /cache/sha256:9cbf21e56fa11e140e68f95fb9d855d7b5a6468e1ccebb647c5078d169eed9d2: no such file or directory
DEBU[0000] Image pingidentity/pingfederate not found in cache
INFO[0000] Retrieving image manifest pingidentity/pingfederate
INFO[0000] Retrieving image pingidentity/pingfederate
INFO[0000] Built cross stage deps: map[]
INFO[0000] Retrieving image manifest pingidentity/pingfederate
INFO[0000] Retrieving image pingidentity/pingfederate
DEBU[0001] No file found for cache key sha256:9cbf21e56fa11e140e68f95fb9d855d7b5a6468e1ccebb647c5078d169eed9d2 stat /cache/sha256:9cbf21e56fa11e140e68f95fb9d855d7b5a6468e1ccebb647c5078d169eed9d2: no such file or directory
DEBU[0001] Image pingidentity/pingfederate not found in cache
INFO[0001] Retrieving image manifest pingidentity/pingfederate
INFO[0001] Retrieving image pingidentity/pingfederate
INFO[0002] Executing 0 build triggers
INFO[0002] Unpacking rootfs as cmd COPY test.lic test.lic requires it.
DEBU[0002] Mounted directories: [{/kaniko false} {/etc/mtab false} {/tmp/apt-key-gpghome true} {/var/run false} {/proc false} {/dev false} {/dev/pts false} {/sys false} {/sys/fs/cgroup false} {/sys/fs/cgroup/systemd false} {/sys/fs/cgroup/perf_event false} {/sys/fs/cgroup/net_cls,net_prio false} {/sys/fs/cgroup/devices false} {/sys/fs/cgroup/pids false} {/sys/fs/cgroup/cpu,cpuacct false} {/sys/fs/cgroup/cpuset false} {/sys/fs/cgroup/hugetlb false} {/sys/fs/cgroup/memory false} {/sys/fs/cgroup/freezer false} {/sys/fs/cgroup/blkio false} {/dev/mqueue false} {/dev/shm false} {/workspace false} {/etc/resolv.conf false} {/etc/hostname false} {/etc/hosts false} {/root/.config/gcloud false} {/proc/bus false} {/proc/fs false} {/proc/irq false} {/proc/sys false} {/proc/sysrq-trigger false} {/proc/acpi false} {/proc/kcore false} {/proc/keys false} {/proc/latency_stats false} {/proc/timer_list false} {/proc/sched_debug false} {/sys/firmware false}]
DEBU[0002] Whiting out /opt/.wh..wh..opq
DEBU[0002] not including whiteout files
DEBU[0002] Whiting out /bin/.wh..wh..opq
DEBU[0002] not including whiteout files
DEBU[0002] Not adding /dev because it is ignored
DEBU[0002] Whiting out /dev/.wh..wh..opq
DEBU[0002] not including whiteout files
DEBU[0002] Whiting out /etc/.wh..wh..opq
DEBU[0002] not including whiteout files
DEBU[0002] Not adding /etc/hostname because it is ignored
DEBU[0002] Not adding /etc/hosts because it is ignored
DEBU[0002] Not adding /etc/mtab because it is ignored
DEBU[0002] Whiting out /home/.wh..wh..opq
DEBU[0002] not including whiteout files
DEBU[0002] Whiting out /lib/.wh..wh..opq
DEBU[0002] not including whiteout files
DEBU[0003] Whiting out /media/.wh..wh..opq
DEBU[0003] not including whiteout files
DEBU[0003] Whiting out /mnt/.wh..wh..opq
DEBU[0003] not including whiteout files
DEBU[0003] Whiting out /opt/backup/.wh..wh..opq
DEBU[0003] not including whiteout files
DEBU[0003] Whiting out /opt/in/.wh..wh..opq
DEBU[0003] not including whiteout files
DEBU[0003] Whiting out /opt/java/.wh..wh..opq
DEBU[0003] not including whiteout files
DEBU[0005] Whiting out /opt/logs/.wh..wh..opq
DEBU[0005] not including whiteout files
DEBU[0005] Whiting out /opt/out/.wh..wh..opq
DEBU[0005] not including whiteout files
DEBU[0005] Whiting out /opt/server/.wh..wh..opq
DEBU[0005] not including whiteout files
DEBU[0012] Whiting out /opt/staging/.wh..wh..opq
DEBU[0012] not including whiteout files
DEBU[0012] Not adding /proc because it is ignored
DEBU[0012] Whiting out /proc/.wh..wh..opq
DEBU[0012] not including whiteout files
DEBU[0012] Whiting out /root/.wh..wh..opq
DEBU[0012] not including whiteout files
DEBU[0012] Whiting out /run/.wh..wh..opq
DEBU[0012] not including whiteout files
DEBU[0012] Whiting out /sbin/.wh..wh..opq
DEBU[0012] not including whiteout files
DEBU[0012] Whiting out /srv/.wh..wh..opq
DEBU[0012] not including whiteout files
DEBU[0012] Not adding /sys because it is ignored
DEBU[0012] Whiting out /sys/.wh..wh..opq
error building image: error building stage: failed to get filesystem from image: removing whiteout sys/.wh..wh..opq: unlinkat /sys/.wh..opq: read-only file system

- This is working if doing via Docker build
> docker build -t ping:1 .
Sending build context to Docker daemon  24.06kB
Step 1/2 : FROM pingidentity/pingfederate
 ---> 898156c692f6
Step 2/2 : COPY test.lic test.lic
 ---> Using cache
 ---> 34b88c8834fe
Successfully built 34b88c8834fe
Successfully tagged ping:1

 | **Description** | **Yes/No** |
 |----------------|---------------|
 | Please check if this a new feature you are proposing        | No |
 | Please check if the build works in docker but not in kaniko | Yes | 
 | Please check if this error is seen when you use `--cache` flag | Yes |
 | Please check if your dockerfile is a multistage dockerfile | No |

### LINKED ISSUE #1944: Missing symlink in the images built using Kaniko
**Actual behavior**

I am attempting to migrate the image build process for our micro services from docker to kaniko and have realized that the kaniko built images fails to create files/symlinks generated by the OS packages.   

**Expected behavior**

The image built from a `Dockerfile` using kaniko should the same as the ones built using Docker, it must include all files including symlinks.

**To Reproduce**
1. Use the following `Dockerfile` and build an image using kaniko and another one using docker, push both images to repo.
```
FROM python:3.10.2-slim

RUN apt-get update && apt-get install -y \
  libglib2.0-0 \
  libgl1-mesa-glx \
  && rm -rf /var/lib/apt/lists/*
```

2. Try running the images to experience the difference. The package `libxdmcp6:amd64` generates a symlink `/usr/lib/x86_64-linux-gnu/libXdmcp.so.6` pointing to `/usr/lib/x86_64-linux-gnu/libXdmcp.so.6.0.0` which is missing in the image built using Kaniko and it **does exist** in the image built using Docker as shown below.

#### Kaniko
```
docker run --rm -it prasadzone/pythonapp:kaniko bash
root@3f1505cd4cee:/# dpkg -L libxdmcp6:amd64
/.
/usr
/usr/lib
/usr/lib/x86_64-linux-gnu
/usr/lib/x86_64-linux-gnu/libXdmcp.so.6.0.0
/usr/share
/usr/share/doc
/usr/share/doc/libxdmcp6
/usr/share/doc/libxdmcp6/changelog.Debian.gz
/usr/share/doc/libxdmcp6/changelog.gz
/usr/share/doc/libxdmcp6/copyright
/usr/lib/x86_64-linux-gnu/libXdmcp.so.6
root@3f1505cd4cee:/# ls -la /usr/lib/x86_64-linux-gnu | grep libXdmcp
-rw-r--r--  1 root root    22728 Mar  2  2017 libXdmcp.so.6.0.0
root@3f1505cd4cee:/# 
```
#### Docker
```
$ docker run --rm -it prasadzone/pythonapp:docker bash
root@864bf398734b:/# dpkg -L libxdmcp6:amd64
/.
/usr
/usr/lib
/usr/lib/x86_64-linux-gnu
/usr/lib/x86_64-linux-gnu/libXdmcp.so.6.0.0
/usr/share
/usr/share/doc
/usr/share/doc/libxdmcp6
/usr/share/doc/libxdmcp6/changelog.Debian.gz
/usr/share/doc/libxdmcp6/changelog.gz
/usr/share/doc/libxdmcp6/copyright
/usr/lib/x86_64-linux-gnu/libXdmcp.so.6
root@864bf398734b:/# ls -la /usr/lib/x86_64-linux-gnu | grep libXdmcp
lrwxrwxrwx  1 root root       17 Mar  2  2017 libXdmcp.so.6 -> libXdmcp.so.6.0.0
-rw-r--r--  1 root root    22728 Mar  2  2017 libXdmcp.so.6.0.0
```

This deviation from docker image is a show stopper as the applications depending on this library will fail to start if we use kaniko based images (will get errors similar to below), the app works fine if we use the image built with docker. 

```
  File "/usr/local/lib/python3.10/site-packages/cv2/__init__.py", line 8, in <module>
    from .cv2 import *
ImportError: libXdmcp.so.6: cannot open shared object file: No such file or directory
```

**Additional Information**
 - Dockerfile (Please refer to the reproduce steps)

 **Triage Notes for the Maintainers**
 

 
 | **Description** | **Yes/No** |
 |----------------|---------------|
 | Please check if this a new feature you are proposing        | <ul><li>- [ ] </li></ul>|
 | Please check if the build works in docker but not in kaniko | <ul><li>- [x] </li></ul>| 
 | Please check if this error is seen when you use `--cache` flag | <ul><li>- [ ] </li></ul>|
 | Please check if your dockerfile is a multistage dockerfile | <ul><li>- [ ] </li></ul>|

### PR DESCRIPTION
**Pull request by gabyx**
- Added a test.

Fixes #1420
Fixes #1944

**Description**

Flatten function is still wrong.

**Submitter Checklist**

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [unit tests](../DEVELOPMENT.md#creating-a-pr)
- [x] Adds integration tests if needed.

_See [the contribution guide](../CONTRIBUTING.md) for more details._

**Reviewer Notes**

- [ ] The code flow looks good. 
- [ ] Unit tests and or integration tests added.

### CODE REVIEW

**Comment by gabyx** [comment]
My observation is the following:

I have made the following (maybe trivial) observations: `LayeredMap.layers` never contains any `.wh.*` whiteout files. This is strange because I assumed that, and the function `getFlattenedPaths` was probably containing old code which did nothing.

Instead I hardly guess
- `LayeredMap.layers` contains the added files
- `LayeredMap.whiteouts` contains the deleted files.

However `LayeredMap.getFlattenedPaths` needs to take these two maps into account, otherwise
we (`main`) we are passing only the accumulated `added` files to the `util.WalkFS` function which is totaly wrong.

**Comment by gabyx** [comment]
The other thing is, that this improvements do not yet resolve the symlink bug I see over multiple stages:

```dockerfile
FROM base as A
...
RUN echo "DEBUGGINGCHECK:" && ls -al /usr/lib/x86_64-linux-gnu/libbsd.so.0

FROM A as End
RUN echo "DEBUGGINGCHECK:" && ls -al /usr/lib/x86_64-linux-gnu/libbsd.so.0 /// FAILING HERE
```

Seems that extracting the stage `0` is somehow flawed and does not produce the symlink.

**Comment by gabyx** [comment]
@tejal29 Could you have a look at this changes that would be great!

@imjasonh and me are investigating these changes which fixes some ugly bugs with missing files in images when multiple stages are involved.
The current changes vastly reduce some strange old code in favor of some more clean behavior in `LayeredMap`

**Comment by gabyx** [comment]
@imjasonh: Hurai, finally CI runs through, with minor modifications. IMO only improvements to status quo. Some local test fixes have been applied too to make local testing run through smoothly.

File: pkg/snapshot/layered_map.go [hunk 1]
```diff
@@ -20,129 +20,179 @@ import (
...
+	// Save current state of image
+	l.updateCurrentImage()
+
+	l.adds = append(l.adds, map[string]string{})
+	l.deletes = append(l.deletes, map[string]struct{}{})
+	l.layerHashCache = map[string]string{} // Erase the hash cache for this new layer.
 }
 
-// Key returns a hash for added files
+// Key returns a hash for added and delted files.
```
Comment (imjasonh) [inline pkg/snapshot/layered_map.go:61]:
```suggestion
// Key returns a hash for added and deleted files.
```

**Review by imjasonh — APPROVED** [review]
There's kind of a lot of changes in this PR, but I think I'm following them all, and at this point it's probably easier just to power through than to split it up. In the future though, smaller single-purpose PRs would be great so we can iteratively make small improvements.

**Comment by gabyx** [comment]
@imjasonh Totally agree. I still try to find the bug dockerfile to add as regression test

**Pull request merged**